algebra/src/curves/mod.rs

```diff
@@ -56,274 +23,106 @@ pub mod tests;
-/// in the correct prime order subgroup.
-pub trait ProjectiveCurve:
-    Eq
-    + Sized
-    + ToBytes
-    + FromBytes
-    + Serialize
-    + for<'a> Deserialize<'a>
-    + CanonicalSerialize
-    + CanonicalDeserialize
-    + SemanticallyValid
-    + FromBytesChecked
-    + Copy
-    + Clone
-    + Default
-    + Send
-    + Sync
-    + Hash
-    + Debug
-    + Display
-    + UniformRand
-    + 'static
-    + Neg<Output = Self>
-    + for<'a> Add<&'a Self, Output = Self>
-    + for<'a> Sub<&'a Self, Output = Self>
-    + for<'a> AddAssign<&'a Self>
-    + for<'a> SubAssign<&'a Self>
-{
-    type ScalarField: PrimeField + SquareRootField + Into<<Self::ScalarField as PrimeField>::BigInt>;
-    type BaseField: Field;
-    type Affine: AffineCurve<Projective = Self, ScalarField = Self::ScalarField>;
-
-    /// Returns the additive identity.
-    #[must_use]
-    fn zero() -> Self;
-
-    /// Returns a fixed generator of unknown exponent.
-    #[must_use]
-    fn prime_subgroup_generator() -> Self;
-
-    /// Determines if this point is the point at infinity.
-    #[must_use]
-    fn is_zero(&self) -> bool;
-
-    /// Checks that the current point is on curve and is in the
-    /// prime order subgroup
-    #[must_use]
-    fn group_membership_test(&self) -> bool;
-
-    /// Normalizes a slice of projective elements so that
-    /// conversion to affine is cheap.
-    fn batch_normalization(v: &mut [Self]);
+    fn batch_from_affine<'a>(vec_affine: &'a [Self::AffineRep]) -> Vec<Self> {
+        vec_affine
+            .iter()
+            .map(|&affine| affine.into())
+            .collect::<Vec<_>>()
+    }
 
-    fn batch_normalization_into_affine(mut v: Vec<Self>) -> Vec<Self::Affine> {
-        Self::batch_normalization(v.as_mut_slice());
-        v.into_iter().map(|p| p.into_affine()).collect()
+    fn batch_into_affine<'a>(vec_self: &'a [Self]) -> Result<Vec<Self::AffineRep>, Error> {
+        vec_self
+            .iter()
+            .map(|&projective| projective.into_affine())
+            .collect::<Result<Vec<_>, _>>()
     }
 
-    /// Checks if the point is already "normalized" so that
-    /// cheap affine conversion is possible.
-    #[must_use]
-    fn is_normalized(&self) -> bool;
+    fn add_affine<'a>(&self, other: &'a Self::AffineRep) -> Self;
 
-    /// Doubles this element.
-    #[must_use]
-    fn double(&self) -> Self {
-        let mut copy = *self;
-        copy.double_in_place();
-        copy
-    }
+    fn add_affine_assign<'a>(&mut self, other: &'a Self::AffineRep);
 
-    fn double_in_place(&mut self) -> &mut Self;
+    // TODO: move to group trait?
+    fn mul_bits<S: AsRef<[u64]>>(&self, bits: BitIterator<S>) -> Self;
 
-    /// Adds an affine element to this element.
-    fn add_assign_mixed(&mut self, other: &Self::Affine);
+    fn mul_bits_affine<'a, S: AsRef<[u64]>>(
+        affine: &'a Self::AffineRep,
+        bits: BitIterator<S>,
+    ) -> Self;
```

Comment:
Maybe we can use Rust std Mul traits to be mixed into the Curve trait ?

Reply:
This function is "internal", Mul and MulAssign traits are implemented based on this function

Reply:
If it's internal it seems to me a bit redundant to be put here. Can we find another solution ?